corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -361,7 +362,10 @@ def get_doc(self, doc_id):
             pass
 
     def doc_to_json(self, doc):
-        return doc.to_json(include_attachments=self.include_attachments)
+        try:
+            return doc.to_json(include_attachments=self.include_attachments)
+        except MissingFormXml:
+            return {}
```

Comment:
Hmm, this is a more generic place than I was imagining when discussing #25985. I was expecting this handler to be added to the ES reindexer code, not dbaccessors. There may be other places (other than ES reindexing code) that use this method that will want to handle this error and do something more reasonable with it. 

Silencing the error and returning a blank document does not seem like a good behavior for a generic function like this.

Reply:
This method is actually on the class `FormReindexAccessor`, so I do think it's specific to that use-case, despite being in `dbaccessors.py`

Reply:
although this is also used by `SqlAppFormSubmissionTrackerReindexerFactory`, and theoretically available to future form-based pillows

Reply:
> This method is actually on the class `FormReindexAccessor`

Oh, I didn't realize that, and now I'm less certain.

The problem I have with this change is that it does not actually skip the document, as in an iteration where an invalid item is simply not yielded, and therefore not processed. Instead this returns an empty document which may be interpreted differently in different contexts. Does an empty document mean the document does not exist, or does it mean the document exists but is empty?

Reply:
From a practical standpoint, the issue is that there's a ton of redirection in this code (writing it out to rubber duck).

The logic of fetching the docs and paginating happens in a generic `SqlDocumentProvider` class.  This calls the `FormReindexAccessor.doc_to_json` method on individual docs, but as a transform only - no option to skip.

Later, the `BulkDocProcessor` evaluates the combined iterator I just described, passing each document to `BulkPillowReindexProcessor`, which has access to `xform_pillow_filter` from higher up the stack.  This is where documents are expected to be skipped.

I could probably add a special case where `doc_to_json` can raise some exception caught by the iterator class provided by `SqlDocumentProvider.  That also feels like a hack, though maybe less of one.

Reply:
> add a special case where doc_to_json can raise some exception

Like have a parameter called, say, `raise_on_missing`? That might preserve current behaviour, and allow the function to be used the way we need it here.

> Instead this returns an empty document

What about using `None` as a special return value to indicate that the form XML is missing? It is clear that it's not a valid empty document. Checking for None makes it easy for iterators to skip it. And it's also easy to replace None with an empty document if that's what we want.

Reply:
> From a practical standpoint, the issue is that there's a ton of redirection in this code...

I agree, I always spend a chunk of time trying to untangle that whenever I have to figure out how it works.

> I could probably add a special case where `doc_to_json` can raise some exception caught by the iterator class provided by `SqlDocumentProvider`. That also feels like a hack, though maybe less of one.

This would seem to be in the spirit of the current code structure.

> What about using `None` as a special return value to indicate that the form XML is missing?

Interesting idea. In addition to returning `None`, I think it will also be necessary to make the iteration logic used by `SqlDocumentProvider` (`resumable_sql_model_iterator`) check for and skip `None` in here:

https://github.com/dimagi/commcare-hq/blob/825fdddc3aa1f7cc27531605d6f347f89203bd92/corehq/util/doc_processor/sql.py#L58-L59

I think it would be cleaner to wrap `doc_to_json()` in a try/except with a custom exception.  That's an unmistakable signal:

```python
            try:
                yield reindex_accessor.doc_to_json(doc)
            except MissingDocument:
                pass
```

as opposed to `None` which may originate from a logical coding mistake.

Reply:
> I think it would be cleaner to wrap `doc_to_json()` in a try/except with a custom exception.
> That's an unmistakable signal.

I agree

Reply:
Might also be prudent to call the exception something like `IgnoreDocument` or `SkipDocument, since skipping silently isn't necessarily the obvious way to deal with missing data.  I'll make a PR along those lines and open that for review.